crates/spark/src/services/leaf_optimizer.rs

```diff
@@ -0,0 +1,974 @@
+            if target != batch {
+                swaps.push(SwapPlan {
+                    leaves_to_give: batch.clone(),
+                    leaves_to_receive: target,
+                });
+            }
+            batch.clear();
+        }
+    }
+
+    // Process any remaining leaves
+    if !batch.is_empty() {
+        let batch_sum: u64 = batch.iter().sum();
+        let target = greedy_leaves(batch_sum);
+
+        if target != batch {
+            swaps.push(SwapPlan {
+                leaves_to_give: batch,
+                leaves_to_receive: target,
+            });
+        }
+    }
+
+    swaps
+}
+
+/// Calculates the swaps needed to optimize when minimizing transfer swaps.
+///
+/// Generates swaps that will minimize the probability of needing to swap during a transfer.
+/// Multiple iterations may be required to reach the optimal set.
+fn minimize_transfer_swap(
+    input_leave_amounts: &[u64],
+    multiplicity: u8,
+    max_leaves_per_swap: u32,
+) -> Vec<SwapPlan> {
+    let max_leaves = max_leaves_per_swap as usize;
+
+    let balance: u64 = input_leave_amounts.iter().sum();
+    let optimal_leaves = swap_minimizing_leaves(balance, multiplicity);
+
+    let wallet_counter = count_occurrences(input_leave_amounts);
+    let optimal_counter = count_occurrences(&optimal_leaves);
+
+    let leaves_to_give = subtract_counters(&wallet_counter, &optimal_counter);
+    let leaves_to_receive = subtract_counters(&optimal_counter, &wallet_counter);
+
+    let mut give = counter_to_flat_array(&leaves_to_give);
+    let mut receive = counter_to_flat_array(&leaves_to_receive);
+
+    // Build swaps by balancing give/receive batches
+    let mut swaps = Vec::new();
+    let mut to_give_batch: Vec<u64> = Vec::new();
+    let mut to_receive_batch: Vec<u64> = Vec::new();
+
+    while !give.is_empty() || !receive.is_empty() {
+        let give_sum: u64 = to_give_batch.iter().sum();
+        let receive_sum: u64 = to_receive_batch.iter().sum();
+
+        if give_sum > receive_sum {
+            if receive.is_empty() {
+                break;
+            }
+            to_receive_batch.push(receive.remove(0));
+        } else {
+            if give.is_empty() {
+                break;
+            }
+            to_give_batch.push(give.remove(0));
+        }
+
+        let give_sum: u64 = to_give_batch.iter().sum();
+        let receive_sum: u64 = to_receive_batch.iter().sum();
+
+        if !to_give_batch.is_empty() && !to_receive_batch.is_empty() && give_sum == receive_sum {
+            // Create swap, potentially splitting if too large
+            if to_give_batch.len() > max_leaves {
+                // Split give batch into chunks
+                for chunk in to_give_batch.chunks(max_leaves) {
+                    let chunk_sum: u64 = chunk.iter().sum();
+                    swaps.push(SwapPlan {
+                        leaves_to_give: chunk.to_vec(),
+                        leaves_to_receive: greedy_leaves(chunk_sum),
```

Comment:
Can you explain this step? Above it seems like we have already defined exactly which leaves we want to receive. Here a new set of leaves is created using the greedy algorithm. Shouldn't we be picking leaves from the `to_receive_batch` if possible?

Reply:
IIUC, it will at least move us in the right direction (closer to optimal). The following optimizations should get closer to optimal, even if the current approach didn't reach it due to the imposed maximum leaves per swap.

Also, the max is not configurable on the Breez SDK level, so we should have no issues of someone using some unexpected low number.

Reply:
Yes, I was thinking of that case where there are many small leaves. My thinking was that we'll get fewer leaves, which may help following optimizations reach the optimal set. But I'm not 100% sure about it.

I agree this fallback might not be optimal. Given it is a fallback, I think that's fine. In any case, I've added a TODO so that we don't forget there might be some margin for improvement.